appserver/payara-appserver-modules/healthcheck-checker/src/main/java/fish/payara/healthcheck/mphealth/MicroProfileHealthChecker.java

```diff
@@ -129,111 +135,148 @@ protected HealthCheckResult doCheckInternal() {
-            }
+                return -1;
+            }));
         }
 
         for (InstanceDescriptor instance : payaraMicro.getClusteredPayaras()) {
             String instanceName = instance.getInstanceName();
-            if (usedInstances.contains(instanceName)) {
+            if (tasks.containsKey(instanceName)) {
                 continue;
             }
-
-            Future<?> taskResult = payaraExecutorService.submit(() -> {
-                try {
-
-                    ClusterCommandResult mpHealthConfigResult = configs.get(instance.getMemberUUID()).get();
-                    String values = mpHealthConfigResult.getOutput().split("\n")[1];
-                    Boolean enabled = Boolean.parseBoolean(values.split(" ")[0]);
-                    if (enabled) {
-
-                        String endpoint = values.split(" ", 2)[1].trim();
-
-                        URL usedURL = instance.getApplicationURLS().get(0);
-                        URI remote = new URI(usedURL.getProtocol(), usedURL.getUserInfo(), usedURL.getHost(), usedURL.getPort(), "/" + endpoint, null, null);
-
-                        result.add(pingHealthEndpoint(instanceName, remote));
-
-                    }
-                } catch (InterruptedException | ExecutionException ex) {
-                    serverHttpStatus.put(instanceName, HttpURLConnection.HTTP_CLIENT_TIMEOUT);
-                    LOGGER.log(Level.FINE, "Error processing MP Healthcheck checker", ex);
-                    result.add(new HealthCheckResultEntry(HealthCheckResultStatus.CRITICAL, "UNABLE TO CONNECT - " + ex.toString()));
-                } catch (URISyntaxException ex) {
-                    serverHttpStatus.put(instanceName, 420);
-                    result.add(new HealthCheckResultEntry(HealthCheckResultStatus.CHECK_ERROR, "INVALID ENDPOINT: " + ex.getInput()));
-                } catch (ProcessingException ex) {
-                    serverHttpStatus.put(instanceName, 420);
-                    LOGGER.log(Level.FINE, "Error sending JAX-RS Request", ex);
-                    result.add(new HealthCheckResultEntry(HealthCheckResultStatus.CRITICAL, "UNABLE TO CONNECT - " + ex.getMessage()));
+            tasks.put(instanceName, payaraExecutorService.submit(() -> {
+                ClusterCommandResult mpHealthConfigResult = configs.get(instance.getMemberUUID()) //
+                        .get(timeoutMillis, MILLISECONDS);
+                String values = mpHealthConfigResult.getOutput().split("\n")[1];
+                Boolean enabled = Boolean.parseBoolean(values.split(" ")[0]);
+                if (enabled) {
+                    String endpoint = values.split(" ", 2)[1].trim();
+                    return pingHealthEndpoint(buildURI(instance, endpoint));
                 }
-
-            });
-
-            try {
-                taskResult.get(options.getTimeout(), TimeUnit.MILLISECONDS);
```

Comment:
NB: I believe it was unintentional that the future was resolved within the loop essentially making the use of asynchronous computation ineffective. The changed implementation will dispatch work first and later resolve the futures so these can happen in parallel.